### DISCUSSION TITLE
Image.Set does not respect a subImage's boundaries.

### DISCUSSION THREAD

**Discussion by Goof512** [hajimehoshi/ebiten#2140] [Q&A] [answered]
I notice that the Image.Set() method does not respect the boundaries given by a defined subImage. The below is my minimum viable example of what I am seeing, vs what I would expect, based on the documentation:

```go
package main

import (
	"github.com/hajimehoshi/ebiten/v2"
	"image"
	"image/color"
	_ "image/png"
	"log"
)

const (
	screenWidth  = 640
	screenHeight = 480
)

type Game struct {
}

func (g *Game) Update() error {
	return nil
}

func (g *Game) Draw(screen *ebiten.Image) {

	box := ebiten.NewImage(64, 64)
	subBox := box.SubImage(image.Rect(16, 16, 48, 48)).(*ebiten.Image)

	box.Fill(color.RGBA{0x20, 0x00, 0x00, 0xFF})
	//we set the top left corner correctly here:
	box.Set(0, 0, color.RGBA{0xFF, 0x10, 0x10, 0xFF})

	op := &ebiten.DrawImageOptions{}
	op.GeoM.Translate(8, 8)
	screen.DrawImage(box, op)

	subBox.Fill(color.RGBA{0x00, 0x20, 0x00, 0xFF})
	//but here we can't write to that corner directly
	subBox.Set(0, 0, color.RGBA{0x00, 0xFF, 0x90, 0xFF})
	//and this shouldn't work at all. This should be out of bounds for the sub-image
	subBox.Set(40, 40, color.White)

	op = &ebiten.DrawImageOptions{}
	op.GeoM.Translate(8, 8 + 64 + 64)
	screen.DrawImage(box, op)
}

func (g *Game) Layout(outsideWidth, outsideHeight int) (int, int) {
	return screenWidth, screenHeight
}

func main() {

	ebiten.SetWindowSize(screenWidth, screenHeight)
	ebiten.SetWindowTitle("Test For Image.Draw")
	if err := ebiten.RunGame(&Game{}); err != nil {
		log.Fatal(err)
	}
}
```

### Visual Examples:
![EbitenSubImage_Result](https://user-images.githubusercontent.com/12751934/173288046-4bc4c777-44fe-4469-800e-f3c8b1d01388.PNG)

Compare with what I would expect:
![EbitenSubImage_Expected](https://user-images.githubusercontent.com/12751934/173288130-d8bb3bf9-6bdb-44d4-91ec-ef106a557fb2.PNG)

Looking at other issues, it seems Ebiten's usage of "subImages" might not be precisely compatible with what I would expect, though a quick alteration of the set function has corrected this behavior, I am not 100% that it would be accepted. I'm prepared to submit a PR if this isn't already a known bug/feature.

**Comment by hajimehoshi** [comment, marked as answer]
Hi,

This seems an expected behavior.

```go
	//but here we can't write to that corner directly
	subBox.Set(0, 0, color.RGBA{0x00, 0xFF, 0x90, 0xFF})
```

As `subBox`'s left-upper position is (16, 16), (0, 0) is out of boundary of `subBox`. So nothing happens.

```go
	//and this shouldn't work at all. This should be out of bounds for the sub-image
	subBox.Set(40, 40, color.White)
```

This should work as (40, 40) is in the boundary (16,16)-(48,48).

This behavior should be the same as the standard library `image`.

**Reply by Goof512** [reply]
Many thanks, my apologies, I hadn't realized exactly how convoluted the go author's subimage schema was.

**Reply by steampoweredtaco** [reply]
I would of made the same bad assumption.

I'd be interested in a maybe a SubView method that would have the expected behavior of the original OP that provides a relative coordinate system to the a defined box/bounds. It is a reasonable use case and a fair way to differ from Go's assumed use of bounds. If it doesn't make it to the library though, it isn't too hard to write the abstraction ourselves.
